### DISCUSSION TITLE
Correct ways to restart dead sessions from live mode and restore side-panel in non-live mode

### DISCUSSION THREAD

**Discussion by yingzhu146** [michaelb/sniprun#175] [Q&A]
Often when using any language in live mode, sniprun will just die without giving me any way to restart it. It's hard for me to produce cases that are reproducible - e.g. below can't reproduce the bug, but it's what caused it. I do face this quite frequently across languages

e.g. 

```clojure
(def var "print")
var
(#'var)
```

somehow made it die. I then `SnipReplMemoryClean`. Didn't help. None of SnipClose, SnipTerminate, SnipReset, toggling Live mode seemed to help. It also doesn't appear to help to SnipTerminate, disable livemode and SnipRun again which I would've possibly expected to restart with a new sniprun process.

I also get this sometimes get issues in non-live mode where the "TerminalWithCode" view becomes unresponsive - closing it seems insurrectable, none of the above commands will close/reopen it.

Closing and reopening is cumbersome but so far the only solution.

As an aside - also don't get sniprun to report errors - e.g.

```clojure
(a)
```

will not report any error, despite  `VirtualTextErr`

<details>
  <summary>my config if relevant</summary>

```lua
M.setup = function()
    require 'sniprun'.setup({
        live_mode_toggle = 'enable',
        selected_interpreters = { 'Python3_original', 'JS_TS_deno', 'Go_original', 'Cpp_original', 'C_original',
            'Bash_original', 'Lua_nvim', 'Rust_original' },
        repl_enable = { 'Python3_original', 'JS_TS_deno', 'Go_original', 'Cpp_original', 'C_original', 'Bash_original', 'Clojure_fifo' },
        repl_disable = {},
        interpreter_options = {
            Python3_original = {
                interpreter = os.getenv("HOME") .. "/mambaforge/bin/python",
                error_truncate = "long"
            },
            C_original = { compiler = "clang" },
            GFM_original = {
                default_filetype = 'bash'
            }
        },
        borders = 'single',
        snipruncolors = {
            SniprunVirtualTextOk  = { bg = "#1b1b20", fg = "#FFFFFF", ctermbg = "Cyan", cterfg = "Black" },
            SniprunFloatingWinOk  = { fg = "#1b1b20", fg = "#FFFFFF", ctermfg = "Cyan" },
            SniprunVirtualTextErr = { bg = "#881515", fg = "#000000", ctermbg = "DarkRed", cterfg = "Black" },
            SniprunFloatingWinErr = { fg = "#881515", ctermfg = "DarkRed" },
        },
        show_no_output = {
            "Classic",
        },
        display = {
            "VirtualTextOk",
            "TerminalWithCode",
            "VirtualTextErr", --# display error results as virtual text
            "TempFloatingWindow", --# display results in a floating window
            "LongTempFloatingWindow", --# same as above, but only long results. To use with VirtualText__
        },
        display_options = {
            terminal_width = 80, --# change the terminal display option width
        },
        inline_messages = 1
    }
    )
```

</details>

**Comment by michaelb** [comment]
Well that's a hell of a report (in the valuable sense).

I'll look into it the best I can, but some things you describe are expressely designed against, meaning that they should _never_ happen. I hope that's just an edge case triggered by your strange config (it can be simplified a lot).

It'll probably help if you can send me a log (you can find the file at ~/.cache/sniprun/sniprun.log or if you're on a Mac (?)  $XDG_CACHE/sniprun/sniprun.log) after a problematic crash.

**Reply by yingzhu146** [reply]
here is the log file:
[tmp.txt](https://github.com/michaelb/sniprun/files/9285603/tmp.txt)

Although  I would assume this no longer contains yesterday's session? It does contain an instance of toggling live view on and off and no longer showing the errors of regular mode anymore

Ok lot's of confounders here I know so apologies (which is why I'm asking a question rather than a bug report - likely just me messing stuff up)

Three things I've noticed

1) I've disabled  "TerminalWithCode" - this seems to have fixed all non-live mode instability issues. I.e. - I can reliably toggle/restart the interpreter and see errors/execution correctly.

2) If I send lines individually, I get the correct display. If I then toggle live mode on, I get the correct live mode display (my understanding is that errors are not reported in live mode - although that'd be a cool feature along request #168 ). If I toggle back to non-live mode, I still only get the preview defined by live mode (line wise)

3) During live mode, SnipTerminate doesn't appear to do anything, while SnipReset appears to irrepairably close Sniprun (I can't line wise execute anymore, and turning off live mode doesn't repair it, and SnipRestart/MemoryReset neither help in live mode or after toggling back live mode)


While we're at my redundant config - how would I set it such that previous virtual line results are maintained on runs (old lines small results, newest line floating window)?

**Reply by yingzhu146** [reply]
and yes M1 Max MBP OSX 12.5

**Reply by yingzhu146** [reply]
Also unrelated, really enjoying the newly added clojure support, perfectly overlapped with me starting to look at clojure 😄

**Comment by michaelb** [comment]
Hey!

It's been a while but I finally found the time!

In order:

<details><summary> a corrected config</summary>

M.setup = function()
    require 'sniprun'.setup({
        live_mode_toggle = 'enable',
        selected_interpreters = { 'JS_TS_deno', 'Lua_nvim'}, -- the other are already the default for their respective filetypes
        repl_enable = { 'Python3_original', 'JS_TS_deno', 'Clojure_fifo' },-- JS-TS-deno  is REPL by default, C&Cpp&Rust&Go don't have REPL mode, and i would disable bash if I were you
        repl_disable = {},
        interpreter_options = {
            Python3_original = {
                interpreter = os.getenv("HOME") .. "/mambaforge/bin/python",
                error_truncate = "long"
            },
            C_original = { compiler = "clang" },
        },
        show_no_output = {
            "Classic", -- needless as it's not chosen as display, and default anyway
        },
        display = {
            "VirtualTextOk",
            "TerminalWithCode",
            "VirtualTextErr", --# display error results as virtual text
            "LongTempFloatingWindow", --# same as above, but only long results. To use with VirtualText__
        },
        display_options = {
            terminal_width = 80, --# change the terminal display option width
        },
        inline_messages = 1 -- are you sure? If you didn't encounter the error described in the readme, you should let it at the default value
    }
    )
</details>


- VirtualText doesn't show errors: unfortunately, unable to reproduce\* (it's normal in live mode though, because a metric ton of error shouldn't be displayed at each keystrokes)
- Live mode issues: unable to reproduce. SnipReset & other commands works\*
- 1.) may be linked to this \*
- 2.) Was a bug, fix will be merged soon
- 3.) Kinda normal for SnipTerminate, that is more geared towards REPL-enabled interpreters. SnipReset while in live mode works as expected on my side

last point about your config. A combination of VirtualText* and (LongTempFloatingWindow OR TempFloatingWindow) is probably what you want.

\* yeah the important thing. Another user reported an issue on their MBP, also ARM. A nasty race condition happened and while I couldn't debug or reproduce, I have a long-shot guess at the cause. This bug could cause total unresponsiveness like you described but is strictly limited to REPL-enabled interpreters, which may discriminate: if you've had issues even with interpreters that had no/disabled REPL, the incoming fix won't cure magically your bug, otherwise, it just may. finger crossed. Said fix will be included in the next release if confirmed working.
